### ISSUE TITLE
degraded image classification compared to pb graph

### ISSUE THREAD

**Issue by eyaler** [tensorflow/hub#211]
I am seeing degraded performance using tf hub image classification as compared to loading the graph from the pb file. Here is a minimal code showing the issue with inception v3. in this example the max probability drops from 79% to 59%. I have seen similar behavior with other images. I am using tensorflow-gpu==1.12.0, tensorflow-hub==0.2.0

`
import numpy as np
import tensorflow as tf
import tensorflow_hub as hub
import cv2

img = cv2.imread(r'../tensorflow/tensorflow/examples/label_image/data/grace_hopper.jpg')[..., ::-1] / 255.0
img = cv2.resize(img, (299, 299))[None, ...]
images = tf.placeholder(tf.float32, (None, 299, 299, 3))

imagenet_module = hub.Module('https://tfhub.dev/google/imagenet/inception_v3/classification/1')
probs = tf.nn.softmax(imagenet_module(images))
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    hub_predictions = probs.eval(session=sess, feed_dict={images:img})[0]

graph = tf.Graph()
graph_def = tf.GraphDef()
with open('../tensorflow/tensorflow/examples/label_image/data/inception_v3_2016_08_28_frozen.pb', "rb") as f:
    graph_def.ParseFromString(f.read())
with graph.as_default():
    tf.import_graph_def(graph_def)

input_op = graph.get_operation_by_name('import/input').outputs[0]
output_op = graph.get_operation_by_name('import/InceptionV3/Predictions/Reshape_1').outputs[0]

with tf.Session(graph=graph) as sess:
    pb_predictions = output_op.eval(session=sess, feed_dict={input_op: img})[0]

print([(ii, hub_predictions[ii]) for ii in np.argsort(hub_predictions)[-5:][::-1]])
#[(653, 0.5871468), (668, 0.04707367), (401, 0.025522923), (458, 0.019224327), (835, 0.019154694)]
print([(ii, pb_predictions[ii]) for ii in np.argsort(pb_predictions)[-5:][::-1]])
#[(653, 0.7925747), (668, 0.025084553), (716, 0.013979144), (440, 0.01067971), (466, 0.009220557)]
`

**Comment by arnoegw** [comment]
Hi eyaler, that's intriguing... What happens if you change pb_predictions by feeding `2.0*img-0.5` instead of `img`?

**Comment by eyaler** [comment]
> 
> 
> Hi eyaler, that's intriguing... What happens if you change pb_predictions by feeding `2.0*img-0.5` instead of `img`?

yup. using (img+1)/2 for hub fixes my issue for inception_v3.

i tried also nesnet_largem pnasnet_large and inception_resnet_v2. in these the transformation makes a slight difference.

i wonder if hub modules could have some input contract to make sure their input is as expected. in the above case a simple check of non-negativity would have captured it, but more generally even if we cannot check, would be good to provide .get_expected_range etc

**Comment by arnoegw** [comment]
Hi @eyaler, thanks for confirming.

For Hub modules, we try to avoid *any* ambiguity of input range by always using the canonical [0,1] range as it comes out of `convert_image_dtype(decode_(...), float32). (Maybe `uint8` would have been even less ambiguous, but that gets in the way of backpropping through a module.)

The frozen graph you are using appears to come from https://github.com/tensorflow/models/tree/master/research/slim ; and to my recollection, they let input ranges vary by model and training regime (usually [-1,+1] for Inception). That would mean the scale was correct for Hub but wrong for Slim. A higher predicted probability on a single, correctly classified example is no proof for better accuracy on a full dataset... But that's beyond Hub technicalities.

**Comment by eyaler** [comment]
but surely here admiralty deems a higher probability score :) anyway, following is a code recreating the example in: [](https://www.tensorflow.org/tutorials/images/image_recognition) and showing the difference with hub. do you think the issue is with the code in [](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/label_image/label_image.py) ?

import numpy as np
import tensorflow as tf
import tensorflow_hub as hub

with tf.Session() as sess:
    file_reader = tf.read_file('../tensorflow/tensorflow/examples/label_image/data/grace_hopper.jpg')
    image_reader = tf.image.decode_jpeg(
        file_reader, channels=3, name="jpeg_reader")
    float_caster = tf.cast(image_reader, tf.float32)
    dims_expander = tf.expand_dims(float_caster, 0)
    resized = tf.image.resize_bilinear(dims_expander, [299,299])
    normalized = tf.divide(tf.subtract(resized, [0]), [255])
    img = sess.run(normalized)

images = tf.placeholder(tf.float32, (None, 299, 299, 3))

imagenet_module = hub.Module('https://tfhub.dev/google/imagenet/inception_v3/classification/1')
probs = tf.nn.softmax(imagenet_module(images))
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    hub_predictions = probs.eval(session=sess, feed_dict={images:img})[0]
    hub_predictions_trans = probs.eval(session=sess, feed_dict={images: img/2+0.5})[0]

graph = tf.Graph()
graph_def = tf.GraphDef()
with open('../tensorflow/tensorflow/examples/label_image/data/inception_v3_2016_08_28_frozen.pb', "rb") as f:
    graph_def.ParseFromString(f.read())
with graph.as_default():
    tf.import_graph_def(graph_def)

    input_op = graph.get_operation_by_name('import/input').outputs[0]
    output_op = graph.get_operation_by_name('import/InceptionV3/Predictions/Reshape_1').outputs[0]

    with tf.Session(graph=graph) as sess:
        pb_predictions = output_op.eval(session=sess, feed_dict={input_op: img})[0]

print(np.min(img),np.max(img))
print([(ii, hub_predictions[ii]) for ii in np.argsort(hub_predictions)[-5:][::-1]])
print([(ii, hub_predictions_trans[ii]) for ii in np.argsort(hub_predictions_trans)[-5:][::-1]])
print([(ii, pb_predictions[ii]) for ii in np.argsort(pb_predictions)[-5:][::-1]])

**Comment by eyaler** [comment]
@arnoegw could you please reopen the bug or otherwise confirm that the error is in the tutorial example?

**Comment by arnoegw** [comment]
From what I can see, there is no issue with tensorflow_hub: the Hub module wraps the SLIM checkpoint with the intended transformation of input range. That's why I closed this issue.

Your findings indicate there might be a range error in this old tensorflow.org example (if not in the SLIM frozen graph itself), but I haven't looked into that. This example is so outdated in many ways, it should be replaced entirely, and if that happens with TF-Hub, the concern about manually adjusting input ranges will go away.

**Comment by arnoegw** [comment]
I'm sorry I have no better message for you. Thank you nonetheless for bringing this up! The effect is counter-intuitive enough to warrant double-checking.
